File: src/main/java/org/moinex/ui/dialog/investment/EditInvestmentTargetController.java
/*
 * Filename: EditInvestmentTargetController.java
 * Created on: January  2, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.investment;

import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import javafx.fxml.FXML;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import lombok.NoArgsConstructor;
import org.moinex.model.enums.AssetType;
import org.moinex.model.investment.InvestmentTarget;
import org.moinex.service.I18nService;
import org.moinex.service.InvestmentTargetService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

@Controller
@NoArgsConstructor
public class EditInvestmentTargetController {

    @FXML private VBox targetsContainer;
    @FXML private Label totalPercentageLabel;
    @FXML private Label validationLabel;

    private InvestmentTargetService investmentTargetService;
    private I18nService i18nService;

    private List<TargetRow> targetRows = new ArrayList<>();

    @Autowired
    public EditInvestmentTargetController(
            InvestmentTargetService investmentTargetService, I18nService i18nService) {
        this.investmentTargetService = investmentTargetService;
        this.i18nService = i18nService;
    }

    @FXML
    public void initialize() {
        loadInvestmentTargets();
        setupListeners();
    }

    private void loadInvestmentTargets() {
        targetsContainer.getChildren().clear();
        targetRows.clear();

        for (AssetType assetType : AssetType.values()) {
            try {
                InvestmentTarget target = investmentTargetService.getTargetByType(assetType);
                addTargetRow(assetType, target.getTargetPercentage());
            } catch (EntityNotFoundException e) {
                addTargetRow(assetType, BigDecimal.ZERO);
            }
        }

        updateTotalPercentage();
    }

    private void addTargetRow(AssetType assetType, BigDecimal targetPercentage) {
        HBox row = new HBox(10.0);
        row.setAlignment(Pos.CENTER_LEFT);

        Label typeLabel = new Label(UIUtils.translateAssetType(assetType, i18nService));
        typeLabel.setMinWidth(200.0);
        typeLabel.setStyle("-fx-font-weight: bold;");

        TextField percentageField = new TextField();
        percentageField.setText(targetPercentage.toString());
        percentageField.setPromptText("0.00");
        percentageField.setPrefWidth(100.0);

        Label percentLabel = new Label("%");

        row.getChildren().addAll(typeLabel, percentageField, percentLabel);
        targetsContainer.getChildren().add(row);

        TargetRow targetRow = new TargetRow(assetType, percentageField);
        targetRows.add(targetRow);

        percentageField
                .textProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (!newValue.matches(Constants.PERCENTAGE_REGEX)) {
                                percentageField.setText(oldValue);
                            } else {
                                updateTotalPercentage();
                            }
                        });
    }

    private void setupListeners() {
        updateTotalPercentage();
    }

    private void updateTotalPercentage() {
        java.util.Map<AssetType, BigDecimal> targets = new java.util.HashMap<>();

        for (TargetRow row : targetRows) {
            try {
                String text = row.percentageField.getText().trim();
                BigDecimal value = text.isEmpty() ? BigDecimal.ZERO : new BigDecimal(text);
                targets.put(row.assetType, value);
            } catch (NumberFormatException ignored) {
            }
        }

        InvestmentTargetService.Companion.ValidationResult validationResult =
                investmentTargetService.validate(targets);

        totalPercentageLabel.setText(
                UIUtils.formatPercentage(validationResult.getTotal(), i18nService));

        if (validationResult.isValid()
                && validationResult.getTotal().compareTo(new BigDecimal("100")) == 0) {
            totalPercentageLabel.setStyle("-fx-font-weight: bold; -fx-text-fill: green;");
            validationLabel.setVisible(false);
        } else {
            totalPercentageLabel.setStyle("-fx-font-weight: bold; -fx-text-fill: red;");

            if (!validationResult.isValid()) {
                validationLabel.setText(validationResult.getErrors().getFirst());
            } else if (validationResult.getTotal().compareTo(new BigDecimal("100")) != 0) {
                validationLabel.setText(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .INVESTMENT_DIALOG_TOTAL_PERCENTAGE_VALIDATION));
            }
            validationLabel.setVisible(true);
        }
    }

    @FXML
    private void handleSave() {
        java.util.Map<AssetType, BigDecimal> targets = new java.util.HashMap<>();

        for (TargetRow row : targetRows) {
            try {
                String text = row.percentageField.getText().trim();
                BigDecimal percentage = text.isEmpty() ? BigDecimal.ZERO : new BigDecimal(text);
                targets.put(row.assetType, percentage);
            } catch (NumberFormatException e) {
                WindowUtils.showErrorDialog(
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .INVESTMENT_DIALOG_INVALID_PERCENTAGE_TITLE),
                        i18nService.tr(
                                Constants.TranslationKeys
                                        .INVESTMENT_DIALOG_INVALID_PERCENTAGE_MESSAGE));
                return;
            }
        }

        try {
            investmentTargetService.saveTargets(targets);

            WindowUtils.showSuccessDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_TARGET_UPDATED_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys.INVESTMENT_DIALOG_TARGET_UPDATED_MESSAGE));

            targetsContainer.getScene().getWindow().hide();
        } catch (IllegalStateException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .INVESTMENT_DIALOG_TOTAL_PERCENTAGE_VALIDATION_TITLE),
                    e.getMessage());
        } catch (Exception e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .INVESTMENT_DIALOG_ERROR_UPDATING_TARGET_TITLE),
                    e.getMessage());
        }
    }

    @FXML
    private void handleCancel() {
        targetsContainer.getScene().getWindow().hide();
    }

    private static class TargetRow {
        AssetType assetType;
        TextField percentageField;

        TargetRow(AssetType assetType, TextField percentageField) {
            this.assetType = assetType;
            this.percentageField = percentageField;
        }
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/investment/update/EditInvestmentTargetController.kt
/*
 * Filename: EditInvestmentTargetController.kt (original filename: EditInvestmentTargetController.java)
 * Created on: January  2, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 20/03/2026
 */

package org.moinex.ui.dialog.investment.update

import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.geometry.Pos
import javafx.scene.control.Label
import javafx.scene.control.TextField
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import org.moinex.common.constants.Constants
import org.moinex.common.constants.TranslationKeys
import org.moinex.common.extension.isEqual
import org.moinex.common.extension.isNotEqual
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.enums.AssetType
import org.moinex.service.PreferencesService
import org.moinex.service.investment.InvestmentTargetService
import org.springframework.stereotype.Controller
import java.math.BigDecimal

@Controller
class EditInvestmentTargetController(
    private val investmentTargetService: InvestmentTargetService,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var targetsContainer: VBox

    @FXML
    private lateinit var totalPercentageLabel: Label

    @FXML
    private lateinit var validationLabel: Label

    private val targetRows = mutableListOf<TargetRow>()

    companion object {
        private data class TargetRow(
            val assetType: AssetType,
            val percentageField: TextField,
        )

        private const val VALID_TARGET_STYLE = "-fx-font-weight: bold; -fx-text-fill: green;"
        private const val INVALID_TARGET_STYLE = "-fx-font-weight: bold; -fx-text-fill: red;"
        private const val TYPE_LABEL_STYLE = "-fx-font-weight: bold;"
    }

    @FXML
    fun initialize() {
        loadInvestmentTargets()
        setupListeners()
    }

    private fun loadInvestmentTargets() {
        targetsContainer.children.clear()
        targetRows.clear()

        AssetType.entries.forEach { assetType ->
            val targetPercentage =
                runCatching {
                    investmentTargetService.getTargetByType(assetType).targetPercentage
                }.getOrElse {
                    if (it is EntityNotFoundException) BigDecimal.ZERO else throw it
                }

            addTargetRow(assetType, targetPercentage)
        }

        updateTotalPercentage()
    }

    private fun addTargetRow(
        assetType: AssetType,
        targetPercentage: BigDecimal,
    ) {
        val row =
            HBox(10.0).apply {
                alignment = Pos.CENTER_LEFT
            }

        val typeLabel =
            Label(UIUtils.translateAssetType(assetType)).apply {
                minWidth = 200.0
                style = TYPE_LABEL_STYLE
            }

        val percentageField =
            TextField().apply {
                text = targetPercentage.toString()
                promptText = "0.00"
                prefWidth = 100.0
            }

        val percentLabel = Label("%")

        row.children.addAll(typeLabel, percentageField, percentLabel)
        targetsContainer.children.add(row)

        val targetRow = TargetRow(assetType, percentageField)
        targetRows.add(targetRow)

        percentageField.textProperty().addListener { _, oldValue, newValue ->
            if (!newValue.matches(Regex(Constants.PERCENTAGE_REGEX))) {
                percentageField.text = oldValue
            } else {
                updateTotalPercentage()
            }
        }
    }

    private fun setupListeners() {
        updateTotalPercentage()
    }

    private fun updateTotalPercentage() {
        val targets = mutableMapOf<AssetType, BigDecimal>()

        targetRows.forEach { row ->
            runCatching {
                val text = row.percentageField.text.trim()
                val value = if (text.isEmpty()) BigDecimal.ZERO else text.toBigDecimal()
                targets[row.assetType] = value
            }
        }

        val validationResult = investmentTargetService.validate(targets)

        totalPercentageLabel.text = UIUtils.formatPercentage(validationResult.total)

        if (validationResult.isValid && validationResult.total.isEqual(100)) {
            totalPercentageLabel.style = VALID_TARGET_STYLE
            validationLabel.isVisible = false
        } else {
            totalPercentageLabel.style = INVALID_TARGET_STYLE

            validationLabel.text =
                when {
                    !validationResult.isValid -> validationResult.errors.first()
                    validationResult.total.isNotEqual(100) ->
                        preferencesService.translate(
                            TranslationKeys.INVESTMENT_DIALOG_TOTAL_PERCENTAGE_VALIDATION,
                        )
                    else -> ""
                }
            validationLabel.isVisible = true
        }
    }

    @FXML
    private fun handleSave() {
        val targets = mutableMapOf<AssetType, BigDecimal>()

        for (row in targetRows) {
            runCatching {
                val text = row.percentageField.text.trim()
                val percentage = if (text.isEmpty()) BigDecimal.ZERO else text.toBigDecimal()
                targets[row.assetType] = percentage
            }.onFailure {
                if (it is NumberFormatException) {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_INVALID_PERCENTAGE_TITLE),
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_INVALID_PERCENTAGE_MESSAGE),
                    )
                    return
                }
            }
        }

        runCatching {
            investmentTargetService.saveTargets(targets)

            WindowUtils.showSuccessDialog(
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_TARGET_UPDATED_TITLE),
                preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_TARGET_UPDATED_MESSAGE),
            )

            targetsContainer.scene.window.hide()
        }.onFailure { e ->
            when (e) {
                is IllegalStateException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(
                            TranslationKeys.INVESTMENT_DIALOG_TOTAL_PERCENTAGE_VALIDATION_TITLE,
                        ),
                        e.message ?: "",
                    )
                }
                else -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(TranslationKeys.INVESTMENT_DIALOG_ERROR_UPDATING_TARGET_TITLE),
                        e.message ?: "Unknown error",
                    )
                }
            }
        }
    }

    @FXML
    private fun handleCancel() {
        targetsContainer.scene.window.hide()
    }
}
